### DISCUSSION TITLE
hostname with underscore cannot be resolved

### DISCUSSION THREAD

**Discussion by kerkmann** [rabbitmq/rabbitmq-java-client#1229] [Q&A] [answered]
### Describe the bug

The RabbitMQ cannot resolve hostnames with an underscore.

### Reproduction steps

Create a docker container with an underscore in the name. Create an application with the RabbitMQ library and use the container name with the underscore as hostname.

### Expected behavior

Hostnames should be allowed to have underscore in it.

### Additional context

The answer is just `Connection refused`, using a hostname without an underscore is working.

**Comment by lukebakken** [comment]
At the **bare minimum**, you could have told us the following:

* The hostname you're using
* The docker container and version you're using for RabbitMQ
* The version of the Java client you're using, and the version of Java
* The host operating system you're using, and docker version

An even better issue report would have been some code we could compile and run to see this issue. An _even better_ report than that would be a docker compose project that demonstrates the issue.

Instead, you're expecting us to take time to try to reproduce this issue, when you could have made it very easy for us to do so.

Take a second to think about all of the free software you use, and the millions of hours of work that went into creating it, and you can begin to realize why making it easy for free software providers to reproduce issues is the minimum you can do to help out.

**Reply by lukebakken** [reply]
Finally, here is a project that reproduces the issue. It would have been great to get something like this with the issue report -

https://github.com/lukebakken/rabbitmq-java-client-1229/blob/main-underscore/docker-compose.yml#L4

**Comment by lukebakken** [comment]
This appears related, and is not limited to the Java client or RabbitMQ:

https://forums.docker.com/t/underscore-in-domain-names/12584

cc @acogoluegnes

**Comment by lukebakken** [comment, marked as answer]
FYI, underscore is not _technically_ valid in a hostname:

https://www.digicert.com/kb/ssl-support/underscores-not-allowed-in-fqdns.htm#:~:text=Note%3A%20The%20use%20of%20underscores,to%20be%20used%20in%20hostnames.

https://stackoverflow.com/questions/10959757/the-use-of-the-underscore-in-host-names

https://github.com/moby/moby/issues/27447#issuecomment-254283586

**Comment by michaelklishin** [comment]
RabbitMQ Java client does not limit what hostnames are or are not resolved. In fact, it is not involved in the resolution process but allows you to [customize address resolution](https://github.com/rabbitmq/rabbitmq-java-client/blob/e19a3a6745e6d05f9cfecbfdcc7c9a7075ba942a/src/main/java/com/rabbitmq/client/AddressResolver.java#L24) however you please (you just need to return a list of endpoints that are acceptable hostnames for/by the JDK standard library.
